mainwindow.cpp

```diff
@@ -0,0 +1,249 @@
+            QLayoutItem *item = layout->takeAt(0);
+            if(item != nullptr){
+                item->widget()->deleteLater();
+                delete item;
+            }
+        }
+    }
+}
+
+void MainWindow::updateUI(){
+    updateFieldLabels();
+    updateScoresLabels();
+    if(!playAfterrWin){
+        if(game.isWin()){
+            //game.endGame();
+            showMenu(game.isWin());
+            playAfterrWin = true;
+        }
+    }
+    if(game.isLose()){
+        //game.endGame();
+        showMenu(!game.isLose());
+    }
+}
+
+void MainWindow::updateFieldLabels(){
+    int currentSize = fieldGLayout->count();
+    int newSize = game.getSize();
+    if(currentSize/newSize != newSize){
+        cleanLayout(fieldGLayout);
+        setupFieldLabels();
+    }
+    for(int i = 0; i < newSize; i++){
+        for(int j = 0; j < newSize; j++){
+            QLabel *cell = qobject_cast<QLabel*>(fieldGLayout->itemAtPosition(i,j)->widget());
+            cell->setText(QString::number(game.getFieldCells()[i][j].getValue()));
+        }
+    }
+}
+
+void MainWindow::updateScoresLabels(){
+    //Logic for update score label
+    int currrScore = game.getScore().getCurrentScore();
+    int bestScore = game.getScore().getBestScore();
+    currentScoreLabel->setText("  Current score: " + QString::number(currrScore));
+    if(currrScore > bestScore){
+        bestScore = currrScore;
+        bestScoreLabel->setText("  Best score: " + QString::number(bestScore));
+    }
+}
+
+void MainWindow::onStartButtonClicked(){
+    game.startGame();
+    playAfterrWin = false;
+    updateUI();
+    cleanLayout(buttonsLayout);
+    setupRestartButton();
+}
+
+void MainWindow::showMenu(bool isWin){
+    if(isWin){
+        VictoryGameMenu *victoryMenu = new VictoryGameMenu(this);
+
+        connect(victoryMenu, &VictoryGameMenu::startNewGameSignal, this, &MainWindow::onRestartButtonClicked);
+
+        victoryMenu->exec();
+        delete victoryMenu;
+    }else{
+        DefeatGameMenu *defeatMenu = new DefeatGameMenu(this);
+        connect(defeatMenu, &DefeatGameMenu::startNewGameSignal, this, &MainWindow::onRestartButtonClicked);
+
+        defeatMenu->exec();
+        delete defeatMenu;
+    }
+}
+
+void MainWindow::onRestartButtonClicked(){
+    game.endGame();
+    int fieldOfLabelsSize = sqrt(fieldGLayout->count());
+    game.getScore().resetScore();
+    playAfterrWin = false;
+    game.startGame(fieldOfLabelsSize);
+
+    updateUI();
+}
+
+void MainWindow::keyPressEvent(QKeyEvent *event){ //It can be better
+    if(event->key() == Qt::Key_A || event->key() == Qt::Key_Left){
+        game.moveCellsOnField(Convertor::convertSignalToDir("LEFT"));
+    }else if(event->key() == Qt::Key_D || event->key() == Qt::Key_Right){
+        game.moveCellsOnField(Convertor::convertSignalToDir("RIGHT"));
+    }else if(event->key() == Qt::Key_W || event->key() == Qt::Key_Up){
+        game.moveCellsOnField(Convertor::convertSignalToDir("UP"));
```

Comment:
moveCellsOnField is a bad name for this method, why not just move? Or you can even do a 4 methods inside the game to avoid using direction enum here e.g. moveLeft moveRight etc.

Reply:
done